packages/react-components/src/components/InputBoxComponent.tsx

```diff
@@ -104,24 +171,317 @@ export const InputBoxComponent = (props: InputBoxComponentProps): JSX.Element =>
+      // Both e.currentTarget.selectionStart !== selectionStartValue and e.currentTarget.selectionEnd !== selectionEndValue can be true when a user selects a text by double click
+      if (event.currentTarget.selectionStart !== null && event.currentTarget.selectionStart !== selectionStartValue) {
+        // the selection start is changed
+        const mentionTag = findMentionTagForSelection(tagsValue, event.currentTarget.selectionStart);
+        if (mentionTag !== undefined && mentionTag.plainTextBeginIndex !== undefined) {
+          //TODO: here it takes -1 when it shouldn't, update selectionstart and end with mouse move and or touch move
+          updatedStartIndex = findNewSelectionIndexForMention(
+            mentionTag,
+            inputTextValue,
+            event.currentTarget.selectionStart,
+            selectionStartValue ?? -1
+          );
+        }
+      }
+      if (event.currentTarget.selectionEnd !== null && event.currentTarget.selectionEnd !== selectionEndValue) {
+        // the selection end is changed
+        const mentionTag = findMentionTagForSelection(tagsValue, event.currentTarget.selectionEnd);
+        if (mentionTag !== undefined && mentionTag.plainTextBeginIndex !== undefined) {
+          //TODO: here it takes -1 when it shouldn't, update selectionstart and end with mouse move and or touch move
+          updatedEndIndex = findNewSelectionIndexForMention(
+            mentionTag,
+            inputTextValue,
+            event.currentTarget.selectionEnd,
+            selectionEndValue ?? -1
+          );
+        }
+      }
+    }
+    // e.currentTarget.selectionDirection should be set to handle shift + arrow keys
+    if (event.currentTarget.selectionDirection === null) {
+      event.currentTarget.setSelectionRange(updatedStartIndex, updatedEndIndex);
+    } else {
+      event.currentTarget.setSelectionRange(updatedStartIndex, updatedEndIndex, event.currentTarget.selectionDirection);
+    }
+    setSelectionStartValue(updatedStartIndex);
+    setSelectionEndValue(updatedEndIndex);
+  };
+
+  const getInputFieldTextValue = (): string => {
+    /* @conditional-compile-remove(mention) */
+    return inputTextValue;
+    return textValue;
+  };
+
   return (
     <Stack className={mergedRootStyle}>
       <div className={mergedTextContainerStyle}>
+        {
+          /* @conditional-compile-remove(mention) */ mentionSuggestions.length > 0 && (
+            <_MentionPopover
+              suggestions={mentionSuggestions}
+              activeSuggestionIndex={activeSuggestionIndex}
+              target={inputBoxRef}
+              targetPositionOffset={caretPosition}
+              onRenderSuggestionItem={mentionLookupOptions?.onRenderSuggestionItem}
+              onSuggestionSelected={onSuggestionSelected}
+              onDismiss={() => {
+                updateMentionSuggestions([]);
+              }}
+            />
+          )
+        }
         <TextField
```

Comment:
I can feel all these new logics could be big enough to be one new component (and glad that we got this huge part working well!)

I am thinking of we create a component named TextFieldWithMention
then do this 
const renderTextField = () => {
/* @conditional-compile-remove(mention) */
return <TextFieldWithMention .../>;
return <Texfileld ... />
}

Then you don't have to conditional compile every single lines of new code, most of new logic will be there without CC, you just need to choose whether to import it or not

It's just a small trick for making the feature easier to be stabilized, and easy to adding new logic
And also, we love smaller files ;D

Reply:
A new refactoring task will be created to handle this functionality